File: util/cpp/unionfind.cc
// Union-Find (disjoint set)
// https://www.youtube.com/watch?v=zV3Ul2pA2Fw

#include <vector>

class UnionFind
{

    std::vector<int> parent;

public:
    UnionFind(int n)
    {
        parent.resize(n, -1);
    }

    int root(int x)
    {
        while (parent[x] >= 0)
        {
            x = parent[x];
        }
        return x;
    }

    int size(int x)
    {
        return -parent[root(x)];
    }

    bool connect(int x, int y)
    {
        if (size(x) < size(y))
            std::swap(x, y);

        int root_x = root(x);
        int root_y = root(y);
        if (root_x == root_y)
            return false;
        parent[root_x] += parent[root_y];
        parent[root_y] = root_x;

        return true;
    }

    bool is_connected(int x, int y)
    {
        return (root(x) == root(y));
    }
};

File: util/rust/unionfind.rs
// Union-Find (disjoint set)
// https://www.youtube.com/watch?v=zV3Ul2pA2Fw

struct UnionFind {
    parent: Vec<i32>,
}


impl UnionFind {
    fn new(n: usize) -> UnionFind {
        UnionFind { parent: vec![-1; n] }
    }

    fn root(&self, mut x: usize) -> usize {
        while self.parent[x as usize] >= 0 {
            x = self.parent[x as usize] as usize;
        }
        x
    }

    fn size(&self, x: usize) -> usize {
        let ans = -self.parent[self.root(x)];
        ans as usize
    }

    fn connect(&mut self, x: usize, y: usize) -> bool {
        if self.size(x) < self.size(y) {
            return self.connect(y, x);
        }

        let root_x = self.root(x);
        let root_y = self.root(y);
        if root_x == root_y {
            return false;
        }

        self.parent[root_x] += self.parent[root_y];
        self.parent[root_y] = root_x as i32;

        return true;
    }

    fn is_connected(&self, x: usize, y: usize) -> bool {
        self.root(x) == self.root(y)
    }
}


fn main() {
    let mut uf = UnionFind::new(10);
    assert_eq!(uf.root(3), 3);
    assert!(!uf.is_connected(1, 3));
    assert!(!uf.is_connected(2, 3));
    assert!(uf.is_connected(3, 3));
    assert!(!uf.is_connected(4, 3));
    assert!(!uf.is_connected(5, 3));
    assert_eq!(uf.size(3), 1);
    uf.connect(1, 2);
    uf.connect(2, 3);
    uf.connect(1, 4);
    assert!(uf.is_connected(1, 3));
    assert!(uf.is_connected(2, 3));
    assert!(uf.is_connected(3, 3));
    assert!(uf.is_connected(4, 3));
    assert!(!uf.is_connected(5, 3));
    assert_eq!(uf.size(3), 4);
}